File: frontend/src/utils/validators.js
// Client-side validation helpers

const NPM_PACKAGE_NAME_RE = /^(@[a-z0-9-~][a-z0-9-._~]*\/)?[a-z0-9-~][a-z0-9-._~]*$/;
const PYPI_PACKAGE_NAME_RE = /^[a-zA-Z0-9\-.*_+!']+$/;
const SEMVER_RE = /^\d+\.\d+\.\d+(-[\w.]+)?(\+[\w.]+)?$/;

export function isValidPackageName(name, registry = 'npm') {
  if (!name || typeof name !== 'string') return false;
  if (name.length > 214) return false;
  
  if (registry === 'pypi') {
    return PYPI_PACKAGE_NAME_RE.test(name);
  }
  
  return NPM_PACKAGE_NAME_RE.test(name);
}

export function isValidVersion(version) {
  if (!version || typeof version !== 'string') return false;
  return SEMVER_RE.test(version);
}

export function isValidUrl(url) {
  try {
    new URL(url);
    return true;
  } catch {
    return false;
  }
}

export function validateAnalyzeForm({ name, version, registry }) {
  const errors = {};

  if (!name?.trim()) {
    errors.name = 'Package name is required';
  } else if (!isValidPackageName(name.trim(), registry)) {
    errors.name = 'Invalid package name format';
  }

  if (!version?.trim()) {
    errors.version = 'Version is required';
  } else if (!isValidVersion(version.trim())) {
    errors.version = 'Invalid version (use semver, e.g. 1.2.3)';
  }

  if (registry && !['npm', 'pypi'].includes(registry)) {
    errors.registry = 'Registry must be npm or pypi';
  }

  return Object.keys(errors).length ? errors : null;
}

File: backend/app/utils/validators.py
"""Input validators - each returns (is_valid, error_msg)."""

import re
from urllib.parse import urlparse

_SEMVER_RE = re.compile(r"^\d+\.\d+\.\d+([.\-+].+)?$")
_PKG_NAME_RE = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9\-_.]*$")
_SUPPORTED_REGISTRIES = {"npm", "pypi"}

Result = tuple[bool, str | None]


def validate_package_name(name: str) -> Result:
    if not name:
        return False, "Package name cannot be empty"
    if len(name) > 255:
        return False, "Package name must be 255 characters or fewer"
    if not _PKG_NAME_RE.match(name):
        return False, "Package name contains invalid characters or starts with a dot/hyphen"
    return True, None


def validate_version_format(version: str) -> Result:
    if not version:
        return False, "Version cannot be empty"
    if not _SEMVER_RE.match(version):
        return False, f"Invalid semver format: {version}"
    return True, None


def validate_registry(registry: str) -> Result:
    if registry.lower() not in _SUPPORTED_REGISTRIES:
        return False, f"Unsupported registry: {registry}. Must be one of {_SUPPORTED_REGISTRIES}"
    return True, None


def validate_risk_score(score: float) -> Result:
    if not 0.0 <= score <= 100.0:
        return False, f"Risk score must be between 0 and 100, got {score}"
    return True, None


def validate_url(url: str) -> Result:
    if not url:
        return False, "URL cannot be empty"
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https"):
        return False, "URL must start with http:// or https://"
    if not parsed.netloc:
        return False, "URL is missing a hostname"
    return True, None


def validate_code_content(code: str) -> Result:
    if not code:
        return False, "Code content cannot be empty"
    if len(code) > 100_000:
        return False, f"Code exceeds max length of 100,000 characters ({len(code)})"
    return True, None

